### PR TITLE
[Jobs] Fix race condition on submitting multiple jobs with the same id

### LINKED ISSUE #31356: [Jobs] Jobs submitted with the same ID in quick succession will both fail, with unfriendly error
### What happened + What you expected to happen

If jobs are submitted with the same ID, the second one will fail with an internal unfriendly error (though it hints at the root cause).  Even if the first job would have succeeded, its status is overwritten with the failed status from the second job.

```
❯ ray job submit --submission-id blah3 --no-wait -- echo hi again & ray job submit --submission-id blah3 --no-wait -- echo hi again

❯     ray job status blah3
Job submission server address: http://127.0.0.1:8265

------------------
Job 'blah3' failed
------------------

Status message: Failed to start Job Supervisor actor: The name _ray_internal_job_actor_blah3 (namespace=SUPERVISOR_ACTOR_RAY_NAMESPACE) is already taken. Please use a different name or get the existing actor using ray.get_actor('_ray_internal_job_actor_blah3', namespace='SUPERVISOR_ACTOR_RAY_NAMESPACE').
```

I would expect that (1) the first command should succeed and its status should reflect that, and (2) the second should fail with `RuntimeError: Job blah3 already exists.`. This currently happens if the first command is given a second or so to run and update its internal `JobInfo`, but this should still happen even if the commands are issued right after one another.

### Versions / Dependencies

master, MacOS, Python 3.8

### Reproduction script

Above

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
In the internal KV store, we store a map of Job IDs to their JobInfo (containing Ray Jobs API metadata).

Previously, when submitting a job, we (1) check if the info for  already exists in the internal KV, and then (2) put the new info and job ID into the internal KV. 

This caused a race condition when two jobs with the same submission_id were submitted within a second or so of each other.  Both jobs would see the info doesn't already exist, so both would try to go ahead with the job submission. This would eventually fail with an unfriendly internal error about named actors (JobSupervisor actor) having the same name.

This PR fixes the race condition by making operations (1) and (2) happen at the same time (this is already supported by `internal_kv_put(... overwrite=False) -> int` which returns the number of keys newly added; this PR just updates the Jobs code to use `overwrite=False` and the return value).

Also adds a unit test which fails without this change.

## Related issue number
Closes #31356 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/job/common.py [hunk 1]
```diff
@@ -189,13 +189,28 @@ def __init__(self, gcs_aio_client: GcsAioClient):
...
+
+        Args:
+            job_id: The job id.
+            job_info: The job info.
+            overwrite: Whether to overwrite the existing job info.
+
+        Returns:
+            The number of keys added. If overwrite is True, this will be 1 if the
+                key was added and 0 if the key was updated. If overwrite is False,
+                this will be 1 if the key was added and 0 if the key already exists.
```
Comment (edoakes) [inline dashboard/modules/job/common.py:205]: should this just be a bool instead of a magic int?

**Review by edoakes — APPROVED** [review]
one nit

File: dashboard/modules/job/common.py (hunk 1 above)
Comment (architkulkarni) [reply dashboard/modules/job/common.py:205]: Thought about this, but I worry that the bool would be interpreted as "successful." The semantics are different from `num_added`

Comment (architkulkarni) [reply dashboard/modules/job/common.py:205]: Don't have a strong preference either way though, if you think the bool wouldn't be unexpected or weird I'll change it

Comment (edoakes) [reply dashboard/modules/job/common.py:205]: I see your point.

I think if you make it a bool and name it something like `new_key_added` in the docstring and callsite that's probably clearest, but also no strong preference. it just took me awhile to grok this docstring and a bool would've been easier

**Review by fishbone — APPROVED** [review]
LG

**Comment by architkulkarni** [comment]
`test_tune_restore_warm_start` unrelated
`runtime_env` conda versions failure unrelated, was fixed on master

**Comment by architkulkarni** [comment]
(cc @gvspraveen )

**Pull request merged**